src/daemon/plugin.c

```diff
@@ -1078,14 +1078,66 @@ static int plugin_compare_read_func(const void *arg0, const void *arg1) {
     return 0;
 } /* int plugin_compare_read_func */
 
+/*
+ * Set the time of the first read according to the settings AlignRead and
+ * Interval in collectd.conf.
+ *
+ * The AlignRead value is in seconds. The value modulo 60 determines the second
+ * of a minute. The integer part of the value divided by 60 determines the
+ * minute of an hour. Optionally, decimal places can be used to set the
+ * fraction of a second (up to milliseconds precision).
+ */
+static void plugin_set_first_read_time(read_func_t *rf) {
+
+  double start = rf->rf_ctx.align_read; /* (default is -1.0) */
+
+  /* if AlignRead is not set, start reading immediately */
+  if (start < 0) {
+    rf->rf_next_read = cdtime();
+  } else {
+    cdtime_t now = cdtime();
+
+    /* get minute in hour and second in minute and millisecond in second */
+    struct timeval tv = CDTIME_T_TO_TIMEVAL(now);
+    uint64_t curr_msec =
+        (((tv.tv_sec / 60) % 60) * 60 // minute of hour in seconds
+         + gmtime(&(tv.tv_sec))->tm_sec) *
+            1000             // add seconds of minute and convert to ms
+        + tv.tv_usec / 1000; // add millisecond of second
+
+    /* DEBUG("%s: current time: %lu sec, %lu ms in minute, start time: %.3lf,
+    interval: %lu", rf->rf_name, tv.tv_sec, curr_msec, start,
+    CDTIME_T_TO_MS(rf->rf_interval)); */
+
+    /* determine delay until next (aligned) read */
+    int64_t sleep_cdt = DOUBLE_TO_CDTIME_T(start) - MS_TO_CDTIME_T(curr_msec);
+    while (sleep_cdt < 0) {
```

Comment:
Why do you use loop here instead of mathematical expression?

Reply:
I replaced the loop with an expression.